### ISSUE TITLE
Add ability to match on local listening address

### ISSUE THREAD

**Issue by speijnik** [containous/traefik#3399]
### Do you want to request a *feature* or report a *bug*?

Feature

### What did you expect to see?

Especially in larger, dynamic environments it might be useful to not only define a Host or HostRegexp matcher, but also match on the local IP address or IP:port combination that is handling the incoming request.
This allows pinning specific IPs to specific backends.

**Comment by dtomcej** [comment]
Hello @speijnik,

We reviewed your issue and your PR, and we are a bit confused.
Right now you ware able to configure an entrypoint with an IP/port combination, and can route traefik through them using entrypoint bindings:

```
[entryPoints]
  [entryPoints.httpInternal]
  address = "127.0.0.1:80"
  [entryPoints.httpExternal]
  address = "24.74.56.55:80"
```
And then wiring your frontends to those entrypoints. Can you elaborate a bit more as to why this would not work in your case?

The only case that I can think of that this would be required is if you were to have traefik bound to all interfaces: `:80`, but then could not tell which interface the request came in on. That being said, the  resolution would be to bind to the interfaces per-IP as demonstrated above.

Thanks!

**Comment by speijnik** [comment]
Hello @dtomcej,

I am happy to elaborate on the use-case and the idea behind it:

Creating multiple entrypoints and binding these entrypoints to specific IP addresses does not work when the IP addresses available on the host Traefik is running on changed dynamically, as the corresponding IP address may not be available (yet).

In our use-case this is caused by using keepalived/VRRP to have IP-level failover for multiple hosts running Traefik. In short this means that an IP address might not be configured when Traefik starts, but may become available anytime while Traefik is running. 

Using the aproach you mentioned would add some complexity, as the Traefik configuration would need to be re-generated and Traefik re-started once our IP configuration changes due to a failover.
Additionally, this likely has negative effects during failover, as Traefik needs to be restarted in this case, creating a short amount of time in which the host that we are failing over to is also unable to process incoming connections.

This also applies for all environments where IP addresses are added and removed dynamically, with the same implications that Traefik would need to be re-configured and restarted everytime the IP configuration changes.

So, in short, binding to `:80` in combination with the LocalIP/LocalAddr matchers in the PR give us the ability to have such a dynamic environment, without having to fully re-configure and restart Traefik everytime the underlying IP configuration changes.
Getting away without having to apply this change to Traefik reduces complexity of the overall system and eliminates (short) downtimes caused by this process.

**Comment by dtomcej** [comment]
Hello @speijnik,

I first off would like to thank you for your time and contribution.

Discussed this again with the team with the information you provided.

Unfortunately the team feels that using layer 4 elements (SourceIP/Port/DestinationIP/etc) to make routing decisions at the layer 7 level is not really needed by the project at this time.

- The first issue is that many orchestration layers, and proxy layers obstruct or rewrite the source/destination portions of request packets, and therefore cannot be relied upon for l7 routing.
- The second issue is that what you are intending to do can be accomplished by using L7 elements such as request headers, etc, which were designed to aid in routing through modern web proxies.
- The third issue is that by accepting this feature, we are going to have to support it, in environments where it is not really a great fit.
- The fourth issue, is that although it makes sense in your use case, we don't feel that this feature will be widely used.

For these reasons, we unfortunately will not be proceeding with adopting this feature. We hope that you understand, and would like to thank you again for your contribution.

If you would like to discuss further, please visit our slack channel (traefik.slack.com) and reach out.

Thanks

**Comment by speijnik** [comment]
Hi @dtomcej,

thank you for the update and for elaborating on this.

I would love to discuss this on your slack channel, but am unable to register (https://traefik.herokuapp.com/ seems to be broken right now). 
Is there any other way of getting myself on board there?

**Comment by mvrhov** [comment]
I would also like to configure traefik listen on all network interfaces. internal and external, so internal services can connect to other internal services via internal network and whomever connects to the same services from outside via external network interface..
